# This file is part of HiGFXback

# requires
REQUIRES="autotools-wrappers-ac-build mpc-build"

pkg-config --exists --print-errors $REQUIRES || exit 1

# configure (ac-2.64)
./configure --disable-libgomp --disable-libitm --disable-libmudflap --disable-libquadmath --disable-libssp --disable-lto --disable-multilib --disable-plugin --disable-static --enable-languages=c++ --with-gxx-include-dir=/include/c++ --with-native-system-header-dir=/include --libexecdir=/lib --prefix=

# build
make

# install
install -d $DESTDIR/bin
install host-*/gcc/cpp $DESTDIR/bin
install host-*/gcc/g++ $DESTDIR/bin
ln -sf g++ $DESTDIR/bin/c++
install host-*/gcc/xgcc $DESTDIR/bin/gcc
ln -sf gcc $DESTDIR/bin/cc
install -d $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/algorithm $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/cctype $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/clocale $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/cstdio $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/cstdlib $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/cstring $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/c_global/cwchar $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/iosfwd $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/list $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/map $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/set $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/string $DESTDIR/include/c++
install -m 644 libstdc++-v3/include/std/utility $DESTDIR/include/c++
install -d $DESTDIR/include/c++/backward
install -m 644 libstdc++-v3/include/backward/binders.h $DESTDIR/include/c++/backward
install -d $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/algorithmfwd.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/allocator.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/basic_string.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/basic_string.tcc $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/char_traits.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/concept_check.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/cpp_type_traits.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/functexcept.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/list.tcc $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/localefwd.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/move.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/ostream_insert.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/postypes.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/range_access.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_algo.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_algobase.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_construct.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_function.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_heap.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_iterator.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_iterator_base_funcs.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_iterator_base_types.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_list.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_map.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_multimap.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_multiset.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_pair.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_relops.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_set.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_tempbuf.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stl_tree.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/include/bits/stringfwd.h $DESTDIR/include/c++/bits
install -d $DESTDIR/include/c++/debug
install -m 644 libstdc++-v3/include/debug/debug.h $DESTDIR/include/c++/debug
install -d $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/include/ext/alloc_traits.h $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/include/ext/atomicity.h $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/include/ext/new_allocator.h $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/include/ext/numeric_traits.h $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/include/ext/type_traits.h $DESTDIR/include/c++/ext
install -m 644 libstdc++-v3/libsupc++/atomic_lockfree_defines.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/libsupc++/cxxabi_forced.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/libsupc++/exception $DESTDIR/include/c++
install -m 644 libstdc++-v3/libsupc++/exception_defines.h $DESTDIR/include/c++/bits
install -m 644 libstdc++-v3/libsupc++/new $DESTDIR/include/c++
install -d $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/atomic_word.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/c++allocator.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/c++config.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/c++locale.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/cpu_defines.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/gthr-default.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/gthr.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -m 644 `ls -d host* | sed 's/host-//'`/libstdc++-v3/include/`ls -d host* | sed 's/host-//'`/bits/os_defines.h $DESTDIR/include/c++/`ls -d host* | sed 's/host-//'`/bits
install -d $DESTDIR/lib
ln -sf ../bin/cpp $DESTDIR/lib/cpp
install host-*/gcc/libgcc_s.so.1 $DESTDIR/lib
ln -sf libgcc_s.so.1 $DESTDIR/lib/libgcc_s.so
install `ls -d host* | sed 's/host-//'`/libstdc++-v3/src/.libs/libstdc++.so.6 $DESTDIR/lib
ln -sf libstdc++.so.6 $DESTDIR/lib/libstdc++.so
install -d $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install host-*/gcc/cc1 $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install host-*/gcc/cc1plus $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/crtbegin.o $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/crtbeginS.o $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/crtend.o $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/crtendS.o $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/crtfastmath.o $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -m 644 host-*/gcc/libgcc.a $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`
install -d $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/cpuid.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/emmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/float.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/ia32intrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/immintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/mm_malloc.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/mmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/pmmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/smmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/stdarg.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/stdbool.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/stddef.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/stdnoreturn.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/tmmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/unwind.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/x86intrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -m 644 host-*/gcc/include/xmmintrin.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include
install -d $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include-fixed
install -m 644 host-*/gcc/include-fixed/limits.h $DESTDIR/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include-fixed

# build.pc
install -d $DESTDIR/share/pkgconfig
cat > $DESTDIR/share/pkgconfig/g++-build.pc << EOF
Name: g++
Version: 4.7.4
Description: GNU Compiler Collection
Requires: $REQUIRES

devel=\\
/bin/c++ \\
/bin/cc \\
/bin/cpp \\
/bin/g++ \\
/bin/gcc \\
/include/c++/algorithm \\
/include/c++/cctype \\
/include/c++/clocale \\
/include/c++/cstdio \\
/include/c++/cstdlib \\
/include/c++/cstring \\
/include/c++/cwchar \\
/include/c++/exception \\
/include/c++/iosfwd \\
/include/c++/list \\
/include/c++/map \\
/include/c++/new \\
/include/c++/set \\
/include/c++/string \\
/include/c++/utility \\
/include/c++/backward/binders.h \\
/include/c++/bits/algorithmfwd.h \\
/include/c++/bits/allocator.h \\
/include/c++/bits/atomic_lockfree_defines.h \\
/include/c++/bits/basic_string.h \\
/include/c++/bits/basic_string.tcc \\
/include/c++/bits/char_traits.h \\
/include/c++/bits/concept_check.h \\
/include/c++/bits/cpp_type_traits.h \\
/include/c++/bits/cxxabi_forced.h \\
/include/c++/bits/exception_defines.h \\
/include/c++/bits/functexcept.h \\
/include/c++/bits/list.tcc \\
/include/c++/bits/localefwd.h \\
/include/c++/bits/move.h \\
/include/c++/bits/ostream_insert.h \\
/include/c++/bits/postypes.h \\
/include/c++/bits/range_access.h \\
/include/c++/bits/stl_algo.h \\
/include/c++/bits/stl_algobase.h \\
/include/c++/bits/stl_construct.h \\
/include/c++/bits/stl_function.h \\
/include/c++/bits/stl_heap.h \\
/include/c++/bits/stl_iterator.h \\
/include/c++/bits/stl_iterator_base_funcs.h \\
/include/c++/bits/stl_iterator_base_types.h \\
/include/c++/bits/stl_list.h \\
/include/c++/bits/stl_map.h \\
/include/c++/bits/stl_multimap.h \\
/include/c++/bits/stl_multiset.h \\
/include/c++/bits/stl_pair.h \\
/include/c++/bits/stl_relops.h \\
/include/c++/bits/stl_set.h \\
/include/c++/bits/stl_tempbuf.h \\
/include/c++/bits/stl_tree.h \\
/include/c++/bits/stringfwd.h \\
/include/c++/debug/debug.h \\
/include/c++/ext/alloc_traits.h \\
/include/c++/ext/atomicity.h \\
/include/c++/ext/new_allocator.h \\
/include/c++/ext/numeric_traits.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/atomic_word.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/c++allocator.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/c++config.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/c++locale.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/cpu_defines.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/gthr-default.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/gthr.h \\
/include/c++/`ls -d host* | sed 's/host-//'`/bits/os_defines.h \\
/lib/cpp \\
/lib/libgcc_s.so \\
/lib/libstdc++.so \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/cc1 \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/cc1plus \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/crtbegin.o \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/crtbeginS.o \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/crtend.o \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/crtendS.o \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/crtfastmath.o \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/libgcc.a \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/cpuid.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/emmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/float.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/ia32intrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/immintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/mm_malloc.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/mmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/pmmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/smmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/stdarg.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/stdbool.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/stddef.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/stdnoreturn.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/tmmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/unwind.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/x86intrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include/xmmintrin.h \\
/lib/gcc/`ls -d host* | sed 's/host-//'`/`cat gcc/BASE-VER`/include-fixed/limits.h

exec=\\
/lib/libgcc_s.so.1 \\
/lib/libstdc++.so.6
EOF

exit

# patch
--- gcc-4.7.4.orig/config.guess
+++ gcc-4.7.4/config.guess
@@ -142,6 +142,31 @@
 
 # Note: order is significant - the case branches are not exclusive.
 
+case "$UNAME_SYSTEM" in
+Linux|GNU|GNU/*)
+	LIBC=unknown
+
+	eval $set_cc_for_build
+	cat <<-EOF > $dummy.c
+	#include <features.h>
+	#if defined(__UCLIBC__)
+	LIBC=uclibc
+	#elif defined(__dietlibc__)
+	LIBC=dietlibc
+	#elif defined(__GLIBC__)
+	LIBC=gnu
+	#endif
+	EOF
+	eval `$CC_FOR_BUILD -E $dummy.c 2>/dev/null | grep '^LIBC' | sed 's, ,,g'`
+
+	if [ "$LIBC" = unknown ] &&
+	   command -v ldd >/dev/null &&
+	   ldd --version 2>&1 | grep -q ^musl; then
+		LIBC=musl
+	fi
+	;;
+esac
+
 case "${UNAME_MACHINE}:${UNAME_SYSTEM}:${UNAME_RELEASE}:${UNAME_VERSION}" in
     *:NetBSD:*:*)
 	# NetBSD (nbsd) targets should (where applicable) match one or
@@ -872,56 +897,48 @@
 	  EV68*) UNAME_MACHINE=alphaev68 ;;
 	esac
 	objdump --private-headers /bin/sh | grep -q ld.so.1
-	if test "$?" = 0 ; then LIBC="libc1" ; else LIBC="" ; fi
-	echo ${UNAME_MACHINE}-unknown-linux-gnu${LIBC}
+	if test "$?" = 0 ; then LIBC="gnulibc1" ; fi
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     arm*:Linux:*:*)
 	eval $set_cc_for_build
 	if echo __ARM_EABI__ | $CC_FOR_BUILD -E - 2>/dev/null \
 	    | grep -q __ARM_EABI__
 	then
-	    echo ${UNAME_MACHINE}-unknown-linux-gnu
+	    echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	else
 	    if echo __ARM_PCS_VFP | $CC_FOR_BUILD -E - 2>/dev/null \
 		| grep -q __ARM_PCS_VFP
 	    then
-		echo ${UNAME_MACHINE}-unknown-linux-gnueabi
+		echo ${UNAME_MACHINE}-unknown-linux-${LIBC}eabi
 	    else
-		echo ${UNAME_MACHINE}-unknown-linux-gnueabihf
+		echo ${UNAME_MACHINE}-unknown-linux-${LIBC}eabihf
 	    fi
 	fi
 	exit ;;
     avr32*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     cris:Linux:*:*)
-	echo cris-axis-linux-gnu
+	echo cris-axis-linux-${LIBC}
 	exit ;;
     crisv32:Linux:*:*)
-	echo crisv32-axis-linux-gnu
+	echo crisv32-axis-linux-${LIBC}
 	exit ;;
     frv:Linux:*:*)
-	echo frv-unknown-linux-gnu
+	echo frv-unknown-linux-${LIBC}
 	exit ;;
     i*86:Linux:*:*)
-	LIBC=gnu
-	eval $set_cc_for_build
-	sed 's/^	//' << EOF >$dummy.c
-	#ifdef __dietlibc__
-	LIBC=dietlibc
-	#endif
-EOF
-	eval `$CC_FOR_BUILD -E $dummy.c 2>/dev/null | grep '^LIBC'`
 	echo "${UNAME_MACHINE}-pc-linux-${LIBC}"
 	exit ;;
     ia64:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     m32r*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     m68*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     mips:Linux:*:* | mips64:Linux:*:*)
 	eval $set_cc_for_build
@@ -940,54 +957,54 @@
 	#endif
 EOF
 	eval `$CC_FOR_BUILD -E $dummy.c 2>/dev/null | grep '^CPU'`
-	test x"${CPU}" != x && { echo "${CPU}-unknown-linux-gnu"; exit; }
+	test x"${CPU}" != x && { echo "${CPU}-unknown-linux-${LIBC}"; exit; }
 	;;
     or32:Linux:*:*)
-	echo or32-unknown-linux-gnu
+	echo or32-unknown-linux-${LIBC}
 	exit ;;
     padre:Linux:*:*)
-	echo sparc-unknown-linux-gnu
+	echo sparc-unknown-linux-${LIBC}
 	exit ;;
     parisc64:Linux:*:* | hppa64:Linux:*:*)
-	echo hppa64-unknown-linux-gnu
+	echo hppa64-unknown-linux-${LIBC}
 	exit ;;
     parisc:Linux:*:* | hppa:Linux:*:*)
 	# Look for CPU level
 	case `grep '^cpu[^a-z]*:' /proc/cpuinfo 2>/dev/null | cut -d' ' -f2` in
-	  PA7*) echo hppa1.1-unknown-linux-gnu ;;
-	  PA8*) echo hppa2.0-unknown-linux-gnu ;;
-	  *)    echo hppa-unknown-linux-gnu ;;
+	  PA7*) echo hppa1.1-unknown-linux-${LIBC} ;;
+	  PA8*) echo hppa2.0-unknown-linux-${LIBC} ;;
+	  *)    echo hppa-unknown-linux-${LIBC} ;;
 	esac
 	exit ;;
     ppc64:Linux:*:*)
-	echo powerpc64-unknown-linux-gnu
+	echo powerpc64-unknown-linux-${LIBC}
 	exit ;;
     ppc:Linux:*:*)
-	echo powerpc-unknown-linux-gnu
+	echo powerpc-unknown-linux-${LIBC}
 	exit ;;
     s390:Linux:*:* | s390x:Linux:*:*)
 	echo ${UNAME_MACHINE}-ibm-linux
 	exit ;;
     sh64*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     sh*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     sparc:Linux:*:* | sparc64:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     tile*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     vax:Linux:*:*)
-	echo ${UNAME_MACHINE}-dec-linux-gnu
+	echo ${UNAME_MACHINE}-dec-linux-${LIBC}
 	exit ;;
     x86_64:Linux:*:*)
-	echo x86_64-unknown-linux-gnu
+	echo x86_64-unknown-linux-${LIBC}
 	exit ;;
     xtensa*:Linux:*:*)
-	echo ${UNAME_MACHINE}-unknown-linux-gnu
+	echo ${UNAME_MACHINE}-unknown-linux-${LIBC}
 	exit ;;
     i*86:DYNIX/ptx:4*:*)
 	# ptx 4.0 does uname -s correctly, with DYNIX/ptx in there.
--- gcc-4.7.4.orig/config.sub
+++ gcc-4.7.4/config.sub
@@ -125,7 +125,7 @@
 maybe_os=`echo $1 | sed 's/^\(.*\)-\([^-]*-[^-]*\)$/\2/'`
 case $maybe_os in
   nto-qnx* | linux-gnu* | linux-android* | linux-dietlibc | linux-newlib* | \
-  linux-uclibc* | uclinux-uclibc* | uclinux-gnu* | kfreebsd*-gnu* | \
+  linux-musl | linux-uclibc* | uclinux-uclibc* | uclinux-gnu* | kfreebsd*-gnu* | \
   knetbsd*-gnu* | netbsd*-gnu* | \
   kopensolaris*-gnu* | \
   storm-chaos* | os2-emx* | rtmk-nova*)
@@ -1345,7 +1345,7 @@
 	      | -chorusos* | -chorusrdb* | -cegcc* \
 	      | -cygwin* | -pe* | -psos* | -moss* | -proelf* | -rtems* \
 	      | -mingw32* | -linux-gnu* | -linux-android* \
-	      | -linux-newlib* | -linux-uclibc* \
+	      | -linux-newlib* | -linux-musl* | -linux-uclibc* \
 	      | -uxpv* | -beos* | -mpeix* | -udk* \
 	      | -interix* | -uwin* | -mks* | -rhapsody* | -darwin* | -opened* \
 	      | -openstep* | -oskit* | -conix* | -pw32* | -nonstopux* \
--- gcc-4.7.4.orig/gcc/config/i386/linux64.h
+++ gcc-4.7.4/gcc/config/i386/linux64.h
@@ -29,5 +29,5 @@
 #define GNU_USER_LINK_EMULATIONX32 "elf32_x86_64"
 
 #define GLIBC_DYNAMIC_LINKER32 "/lib/ld-linux.so.2"
-#define GLIBC_DYNAMIC_LINKER64 "/lib64/ld-linux-x86-64.so.2"
-#define GLIBC_DYNAMIC_LINKERX32 "/libx32/ld-linux-x32.so.2"
+#define GLIBC_DYNAMIC_LINKER64 "/lib/ld-linux-x86-64.so.2"
+#define GLIBC_DYNAMIC_LINKERX32 "/lib/ld-linux-x32.so.2"
--- gcc-4.7.4.orig/gcc/config/i386/t-linux64
+++ gcc-4.7.4/gcc/config/i386/t-linux64
@@ -34,6 +34,6 @@
 comma=,
 MULTILIB_OPTIONS    = $(subst $(comma),/,$(TM_MULTILIB_CONFIG))
 MULTILIB_DIRNAMES   = $(patsubst m%, %, $(subst /, ,$(MULTILIB_OPTIONS)))
-MULTILIB_OSDIRNAMES = m64=../lib64$(call if_multiarch,:x86_64-linux-gnu)
+MULTILIB_OSDIRNAMES = m64=../lib$(call if_multiarch,:x86_64-linux-gnu)
 MULTILIB_OSDIRNAMES+= m32=$(if $(wildcard $(shell echo $(SYSTEM_HEADER_DIR))/../../usr/lib32),../lib32,../lib)$(call if_multiarch,:i386-linux-gnu)
 MULTILIB_OSDIRNAMES+= mx32=../libx32$(call if_multiarch,:x86_64-linux-gnux32)
--- gcc-4.7.4.orig/gcc/config/linux.h
+++ gcc-4.7.4/gcc/config/linux.h
@@ -33,10 +33,12 @@
 #define OPTION_GLIBC  (DEFAULT_LIBC == LIBC_GLIBC)
 #define OPTION_UCLIBC (DEFAULT_LIBC == LIBC_UCLIBC)
 #define OPTION_BIONIC (DEFAULT_LIBC == LIBC_BIONIC)
+#define OPTION_MUSL   (DEFAULT_LIBC == LIBC_MUSL)
 #else
 #define OPTION_GLIBC  (linux_libc == LIBC_GLIBC)
 #define OPTION_UCLIBC (linux_libc == LIBC_UCLIBC)
 #define OPTION_BIONIC (linux_libc == LIBC_BIONIC)
+#define OPTION_MUSL   (linux_libc == LIBC_MUSL)
 #endif
 
 #define GNU_USER_TARGET_OS_CPP_BUILTINS()			\
@@ -51,21 +53,25 @@
     } while (0)
 
 /* Determine which dynamic linker to use depending on whether GLIBC or
-   uClibc or Bionic is the default C library and whether
-   -muclibc or -mglibc or -mbionic has been passed to change the default.  */
+   uClibc or Bionic or musl is the default C library and whether
+   -muclibc or -mglibc or -mbionic or -mmusl has been passed to change
+   the default.  */
 
-#define CHOOSE_DYNAMIC_LINKER1(LIBC1, LIBC2, LIBC3, LD1, LD2, LD3)	\
-  "%{" LIBC2 ":" LD2 ";:%{" LIBC3 ":" LD3 ";:" LD1 "}}"
+#define CHOOSE_DYNAMIC_LINKER1(LIBC1, LIBC2, LIBC3, LIBC4, LD1, LD2, LD3, LD4)	\
+  "%{" LIBC2 ":" LD2 ";:%{" LIBC3 ":" LD3 ";:%{" LIBC4 ":" LD4 ";:" LD1 "}}}"
 
 #if DEFAULT_LIBC == LIBC_GLIBC
-#define CHOOSE_DYNAMIC_LINKER(G, U, B) \
-  CHOOSE_DYNAMIC_LINKER1 ("mglibc", "muclibc", "mbionic", G, U, B)
+#define CHOOSE_DYNAMIC_LINKER(G, U, B, M) \
+  CHOOSE_DYNAMIC_LINKER1 ("mglibc", "muclibc", "mbionic", "mmusl", G, U, B, M)
 #elif DEFAULT_LIBC == LIBC_UCLIBC
-#define CHOOSE_DYNAMIC_LINKER(G, U, B) \
-  CHOOSE_DYNAMIC_LINKER1 ("muclibc", "mglibc", "mbionic", U, G, B)
+#define CHOOSE_DYNAMIC_LINKER(G, U, B, M) \
+  CHOOSE_DYNAMIC_LINKER1 ("muclibc", "mglibc", "mbionic", "mmusl", U, G, B, M)
 #elif DEFAULT_LIBC == LIBC_BIONIC
-#define CHOOSE_DYNAMIC_LINKER(G, U, B) \
-  CHOOSE_DYNAMIC_LINKER1 ("mbionic", "mglibc", "muclibc", B, G, U)
+#define CHOOSE_DYNAMIC_LINKER(G, U, B, M) \
+  CHOOSE_DYNAMIC_LINKER1 ("mbionic", "mglibc", "muclibc", "mmusl", B, G, U, M)
+#elif DEFAULT_LIBC == LIBC_MUSL
+#define CHOOSE_DYNAMIC_LINKER(G, U, B, M) \
+  CHOOSE_DYNAMIC_LINKER1 ("mmusl", "mglibc", "muclibc", "mbionic", M, G, U, B)
 #else
 #error "Unsupported DEFAULT_LIBC"
 #endif /* DEFAULT_LIBC */
@@ -82,19 +88,23 @@
 #define BIONIC_DYNAMIC_LINKER32 "/system/bin/linker"
 #define BIONIC_DYNAMIC_LINKER64 "/system/bin/linker64"
 #define BIONIC_DYNAMIC_LINKERX32 "/system/bin/linkerx32"
+#define MUSL_DYNAMIC_LINKER "/lib/ld-musl-i386.so.1"
+#define MUSL_DYNAMIC_LINKER32 "/lib/ld-musl-i386.so.1"
+#define MUSL_DYNAMIC_LINKER64 "/lib/ld-musl-x86_64.so.1"
+#define MUSL_DYNAMIC_LINKERX32 "/lib/ld-musl-x32.so.1"
 
 #define GNU_USER_DYNAMIC_LINKER						\
   CHOOSE_DYNAMIC_LINKER (GLIBC_DYNAMIC_LINKER, UCLIBC_DYNAMIC_LINKER,	\
-			 BIONIC_DYNAMIC_LINKER)
+			 BIONIC_DYNAMIC_LINKER, MUSL_DYNAMIC_LINKER)
 #define GNU_USER_DYNAMIC_LINKER32					\
   CHOOSE_DYNAMIC_LINKER (GLIBC_DYNAMIC_LINKER32, UCLIBC_DYNAMIC_LINKER32, \
-			 BIONIC_DYNAMIC_LINKER32)
+			 BIONIC_DYNAMIC_LINKER32, MUSL_DYNAMIC_LINKER32)
 #define GNU_USER_DYNAMIC_LINKER64					\
   CHOOSE_DYNAMIC_LINKER (GLIBC_DYNAMIC_LINKER64, UCLIBC_DYNAMIC_LINKER64, \
-			 BIONIC_DYNAMIC_LINKER64)
+			 BIONIC_DYNAMIC_LINKER64, MUSL_DYNAMIC_LINKER64)
 #define GNU_USER_DYNAMIC_LINKERX32					\
   CHOOSE_DYNAMIC_LINKER (GLIBC_DYNAMIC_LINKERX32, UCLIBC_DYNAMIC_LINKERX32, \
-			 BIONIC_DYNAMIC_LINKERX32)
+			 BIONIC_DYNAMIC_LINKERX32, MUSL_DYNAMIC_LINKERX32)
 
 /* Determine whether the entire c99 runtime
    is present in the runtime library.  */
--- gcc-4.7.4.orig/gcc/config/linux.opt
+++ gcc-4.7.4/gcc/config/linux.opt
@@ -28,5 +28,9 @@
 Use GNU C library
 
 muclibc
-Target Report RejectNegative Var(linux_libc,LIBC_UCLIBC) Negative(mbionic)
+Target Report RejectNegative Var(linux_libc,LIBC_UCLIBC) Negative(mmusl)
 Use uClibc C library
+
+mmusl
+Target Report RejectNegative Var(linux_libc,LIBC_MUSL) Negative(mbionic)
+Use musl C library
--- gcc-4.7.4.orig/gcc/config.gcc
+++ gcc-4.7.4/gcc/config.gcc
@@ -522,7 +522,7 @@
 esac
 
 # Common C libraries.
-tm_defines="$tm_defines LIBC_GLIBC=1 LIBC_UCLIBC=2 LIBC_BIONIC=3"
+tm_defines="$tm_defines LIBC_GLIBC=1 LIBC_UCLIBC=2 LIBC_BIONIC=3 LIBC_MUSL=4"
 
 # Common parts for widely ported systems.
 case ${target} in
@@ -625,6 +625,9 @@
     *-*-*uclibc*)
       tm_defines="$tm_defines DEFAULT_LIBC=LIBC_UCLIBC"
       ;;
+    *-*-*musl*)
+      tm_defines="$tm_defines DEFAULT_LIBC=LIBC_MUSL"
+      ;;
     *)
       tm_defines="$tm_defines DEFAULT_LIBC=LIBC_GLIBC"
       ;;
--- gcc-4.7.4.orig/libstdc++-v3/configure.host
+++ gcc-4.7.4/libstdc++-v3/configure.host
@@ -242,6 +242,9 @@
   freebsd*)
     os_include_dir="os/bsd/freebsd"
     ;;
+  linux-musl*)
+    os_include_dir="os/generic"
+    ;;
   gnu* | linux* | kfreebsd*-gnu | knetbsd*-gnu)
     if [ "$uclibc" = "yes" ]; then
       os_include_dir="os/uclibc"

# source
https://gcc.gnu.org/pub/gcc/releases/gcc-4.7.4/gcc-4.7.4.tar.bz2
